Added ppc64le arch migrator for different cuda versions

alhad-deshpande opened the pull request:
Added this migrator for ppc64le arch to build arrow-cpp with CUDA.

<!--
Thank you for your pull request.
Below are a few things we ask you kindly to self-check before getting a review. Remove checks that are not relevant.
-->
Checklist
* [ ] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [ ] Bumped the build number (if the version is unchanged)
* [ ] Reset the build number to `0` (if the version changed)
* [ ] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [ ] Ensured the license file is being packaged.

<!--
Please note any issues this fixes using [closing keywords]( https://help.github.com/articles/closing-issues-using-keywords/ ):
-->

Partially addresses https://github.com/conda-forge/conda-forge-pinning-feedstock/issues/725

<!--
Please add any other relevant info below:
-->

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipe```) and found it was in an excellent condition.

alhad-deshpande commented:
@conda-forge-admin, please ping team

alhad-deshpande commented:
@xhochy 
As suggested by you on https://github.com/conda-forge/arrow-cpp-feedstock/pull/659, raised a PR with migrator for ppc64le to build arrow-cpp with CUDA. Please review the same.

isuruf commented on recipe/migrations/cuda102_110_111_112_115_116_ppc64le.yaml:
```diff
@@ -0,0 +1,68 @@
+  - 11.0                       # [ppc64le]
+  - 11.1                       # [ppc64le]
+  - 11.2                       # [ppc64le]
+  - 11.5                       # [ppc64le]
+  - 11.6                       # [ppc64le]
```
We build with only up to 11.2 and use those for 11.5 for x86

alhad-deshpande replied:
@isuruf 
I want to build CUDA 21.12.00 for ppc64le which requires CUDA 11.5+as per requirements listed in the link - https://github.com/rapidsai/cudf/blob/v21.12.00/CONTRIBUTING.md#setting-up-your-build-environment.
So, I am not sure if building on 11.2 and using those for 11.5 will work. I have tried building CUDF 21.12.00 on CUDA 11.2 locally and it fails due to missing header files/references. 
Can you please let me know if we can add atleast one of those preferably CUDA 11.6?

isuruf replied:
How would you do build cudf for x86_64? It should be the same way.

alhad-deshpande replied:
I have not built cudf on x86_64. But when I install cudatoolkit 11.2 on ppc64le and try to build cudf 21.12.00 then there reference errors.

kkraus14 replied:
I can confirm that cudf requires building with CUDA 11.5 or newer and then can run on 11.0 or newer. It uses new compiler features / support.

isuruf replied:
That's irrelevant. You can use cuda 11.2 compiled dependencies with cudatoolkit 11.6 when you want to build cudf.

alhad-deshpande replied:
@isuruf 
Ok, removed the 11.5 and 11.6 cuda versions. Can you please approve this PR now?

isuruf replied:
No. Please see my other comment.

alhad-deshpande replied:
@isuruf ,
Can you please help in the other comment and let me know how that can be achieved?

isuruf reviewed (changes requested):
You need to make this migrator run only on feedstocks that have run the ppc64le/aarch64 migrator already. There's an option for it, but can't remember

alhad-deshpande commented:
> You need to make this migrator run only on feedstocks that have run the ppc64le/aarch64 migrator already. There's an option for it, but can't remember

@isuruf 
Not sure about how to do this. Can you please help here?

alhad-deshpande commented:
> > You need to make this migrator run only on feedstocks that have run the ppc64le/aarch64 migrator already. There's an option for it, but can't remember
> 
> @isuruf Not sure about how to do this. Can you please help here?

@isuruf ,
Can you please help in this comment and let me know how this can be achieved?

isuruf commented:
Let me ask @conda-forge/bot, who can help.

beckermr commented:
Check the pypy migrators. They use this option.

alhad-deshpande commented:
> 

@beckermr 
Checked the pypy migrators but did not found that option.

@isuruf 
As ppc64le is used as selector in entire file, it will only run if ppc64le is enabled for a package. So can we go ahead with this?

beckermr commented:
See this one. https://github.com/conda-forge/conda-forge-pinning-feedstock/blob/master/recipe/migrations/pypy37-windows.yaml

Look for wait for migrators.

alhad-deshpande commented:
> See this one. https://github.com/conda-forge/conda-forge-pinning-feedstock/blob/master/recipe/migrations/pypy37-windows.yaml
> 
> Look for wait for migrators.

@beckermr 
Thanks for the reference.

@isuruf 
I have added the wait for migrators as cuda110 and cuda111_112 from ppc64le/aarch64 migrators. Please have a look.

isuruf commented:
>  As ppc64le is used as selector in entire file, it will only run if ppc64le is enabled for a package. So can we go ahead with this?

Can you please stop making false statements? Also please be patient. We are all volunteers here.

>  I have added the wait for migrators as cuda110 and cuda111_112 from ppc64le/aarch64 migrators. Please have a look.

No, it should be a wait for the `aarch64andppc64leaddition` migrator.

isuruf commented:
@beckermr, wouldn't this run into the same issue as the flang PR as this migrator will run on all feedstocks that have a C/C++ compiler used?

beckermr commented:
I cannot tell from the yaml. :(

beckermr commented:
As long as override_cbc_keys is set, it will only touch those: 

https://github.com/regro/cf-scripts/blob/master/conda_forge_tick/auto_tick.py#L724

beckermr commented:
So it will hit all cuda feedstocks.

alhad-deshpande commented:
> > As ppc64le is used as selector in entire file, it will only run if ppc64le is enabled for a package. So can we go ahead with this?
> 
> Can you please stop making false statements? Also please be patient. We are all volunteers here.
> 
> > I have added the wait for migrators as cuda110 and cuda111_112 from ppc64le/aarch64 migrators. Please have a look.
> 
> No, it should be a wait for the `aarch64andppc64leaddition` migrator.

Updated the migrator. Please have a look.

alhad-deshpande commented:
> > As ppc64le is used as selector in entire file, it will only run if ppc64le is enabled for a package. So can we go ahead with this?
> 
> Can you please stop making false statements? Also please be patient. We are all volunteers here.
> 
> > I have added the wait for migrators as cuda110 and cuda111_112 from ppc64le/aarch64 migrators. Please have a look.
> 
> 
No, it should be a wait for the `aarch64andppc64leaddition` migrator.



> > > As ppc64le is used as selector in entire file, it will only run if ppc64le is enabled for a package. So can we go ahead with this?
> > 
> > 
> > Can you please stop making false statements? Also please be patient. We are all volunteers here.
> > > I have added the wait for migrators as cuda110 and cuda111_112 from ppc64le/aarch64 migrators. Please have a look.
> > 
> > 
> > No, it should be a wait for the `aarch64andppc64leaddition` migrator.
> 
> 
The intent was not to make any false statement. Apologize for any wrong statement.
Updated the migrator. Please have a look.

isuruf commented on recipe/migrations/cuda102_110_111_112_ppc64le.yaml:
```diff
@@ -0,0 +1,56 @@
+    - aarch64andppc64leaddition
+
+c_compiler_version:            # [ppc64le]
+  - 9                          # [ppc64le]
+  - 9                          # [ppc64le]
+  - 9                          # [ppc64le]
+  - 9                          # [ppc64le]
```
Can you fix the versions to match x86_64?

alhad-deshpande replied:
@isuruf 
Done. Kept 7 for cuda < 11 and 9 for cuda >= 11. Also changed cudnn versions. Please have a look.

isuruf replied:
See https://github.com/conda-forge/conda-forge-pinning-feedstock/blob/master/recipe/conda_build_config.yaml#L14-L17 for the versions that x86_64 uses.

alhad-deshpande replied:
@isuruf 
Fixed the versions to match x86_64. Please have a look.

isuruf commented:
From a technical perspective, this is ready. I have 2 questions to people who are interested.

1. Should we build for all 4 versions (10.2, 11.0, 11.1, 11.2) or just 11.2?
2. Should we bundle aarch64 and ppc64le like we do for non CUDA builds?

@kkraus14, @jakirkham, @jaimergp, @leofang 

For me personally for 1. I think only 11.2 is enough as this is a new platform and we don't usually
build for older versions. For 2. I'm fine with either way.

jaimergp commented:
1) I am fine with 11.2 only on ppc64le. Some packages use emulation to build too, so it makes sense to reduce the matrix whenever possible.

2) No strong preference either.

leofang commented:
1. No strong opinion. CuPy currently still supports 10.2 and will drop it in the next release, but I understand there are several CUDA features requiring 11.2+. 

To @kkraus14's question, Summit is on driver 450.80.02.

2. I'd like to see aarch64 handled if possible. But I don't understand why we have two parallel PRs (this one and #2001) as pointed out by @h-vetinari.

kkraus14 commented:
> To @kkraus14's question, Summit is on driver 450.80.02.

This translates to CUDA 11.0 for the folks following along from home 😉

alhad-deshpande commented:
@isuruf @leofang @kkraus14 @jaimergp 
I need below two arrow packages from feedstock (https://github.com/conda-forge/arrow-cpp-feedstock) with CUDA enabled which currently builds only with CUDA version 10.2 (https://github.com/conda-forge/arrow-cpp-feedstock/blob/main/recipe/meta.yaml). 
 - arrow-cpp-proc[build=cuda]
 - pyarrow==5.0.0[build=*cuda]

Hence, request you to please keep 10.2 CUDA version with 11.2 if we are planning to remove other versions.

kkraus14 commented:
> @isuruf @leofang @kkraus14 @jaimergp 
> I need below two arrow packages from feedstock (https://github.com/conda-forge/arrow-cpp-feedstock) with CUDA enabled which currently builds only with CUDA version 10.2 (https://github.com/conda-forge/arrow-cpp-feedstock/blob/main/recipe/meta.yaml). 
>  - arrow-cpp-proc[build=cuda]
>  - pyarrow==5.0.0[build=*cuda]
> 
> Hence, request you to please keep 10.2 CUDA version with 11.2 if we are planning to remove other versions.

Arrow only uses the CUDA driver and builds with a sufficiently old version that it doesn't pin moving forward so you should be fine regardless.

jakirkham commented:
cc @prashantkhoje

prashantkhoje commented on recipe/migrations/cuda102_110_111_112_ppc64le.yaml:
```diff
@@ -0,0 +1,56 @@
+  primary_key: cuda_compiler_version
+  wait_for_migrators:
+    - aarch64andppc64leaddition
+
```
Are we expecting removal of versions < 11.2?
```suggestion
c_compiler_version:            # [ppc64le]
  - 10                         # [ppc64le]
cxx_compiler_version:          # [ppc64le]
  - 10                         # [ppc64le]
fortran_compiler_version:      # [ppc64le]
  - 10                         # [ppc64le]

cuda_compiler_version:         # [ppc64le]
  - 11.2                       # [ppc64le]

cudnn:                  # [ppc64le]
  - 8                   # [ppc64le]

cdt_name:  # [linux]
  - cos7   # [ppc64le]

docker_image:                                           # [os.environ.get("BUILD_PLATFORM", "").startswith("linux")]
   - quay.io/condaforge/linux-anvil-ppc64le-cuda:11.2   # [ppc64le and os.environ.get("BUILD_PLATFORM") == "linux-ppc64le"]

```

prashantkhoje reviewed (changes requested):
@isuruf @leofang @kkraus14 @jaimergp Are we expecting removal of versions < 11.2 here?

prashantkhoje commented on recipe/migrations/cuda102_110_111_112_ppc64le.yaml:
```diff
@@ -0,0 +1,56 @@
+
+c_compiler_version:            # [ppc64le]
+  - 7                          # [ppc64le]
+  - 9                          # [ppc64le]
+  - 10                         # [ppc64le]
+  - 10                         # [ppc64le]
+cxx_compiler_version:          # [ppc64le]
+  - 7                          # [ppc64le]
+  - 9                          # [ppc64le]
+  - 10                         # [ppc64le]
+  - 10                         # [ppc64le]
+fortran_compiler_version:      # [ppc64le]
+  - 7                          # [ppc64le]
+  - 9                          # [ppc64le]
+  - 10                         # [ppc64le]
+  - 10                         # [ppc64le]
+
+cuda_compiler_version:         # [ppc64le]
+  - 10.2                       # [ppc64le]
+  - 11.0                       # [ppc64le]
+  - 11.1                       # [ppc64le]
+  - 11.2                       # [ppc64le]
+
+cudnn:                  # [ppc64le]
+  - 7                   # [ppc64le]
+  - 8                   # [ppc64le]
+  - 8                   # [ppc64le]
+  - 8                   # [ppc64le]
+
+cdt_name:  # [linux]
+  - cos7   # [ppc64le]
+  - cos7   # [ppc64le]
+  - cos7   # [ppc64le]
+  - cos7   # [ppc64le]
+
+docker_image:                                           # [os.environ.get("BUILD_PLATFORM", "").startswith("linux")]
+   - quay.io/condaforge/linux-anvil-ppc64le-cuda:10.2   # [ppc64le and os.environ.get("BUILD_PLATFORM") == "linux-ppc64le"]
+   - quay.io/condaforge/linux-anvil-ppc64le-cuda:11.0   # [ppc64le and os.environ.get("BUILD_PLATFORM") == "linux-ppc64le"]
+   - quay.io/condaforge/linux-anvil-ppc64le-cuda:11.1   # [ppc64le and os.environ.get("BUILD_PLATFORM") == "linux-ppc64le"]
+   - quay.io/condaforge/linux-anvil-ppc64le-cuda:11.2   # [ppc64le and os.environ.get("BUILD_PLATFORM") == "linux-ppc64le"]
```
Are we expecting removal of versions < 11.2?

```suggestion
c_compiler_version:            # [ppc64le]
  - 10                         # [ppc64le]
cxx_compiler_version:          # [ppc64le]
  - 10                         # [ppc64le]
fortran_compiler_version:      # [ppc64le]
  - 10                         # [ppc64le]

cuda_compiler_version:         # [ppc64le]
  - 11.2                       # [ppc64le]

cudnn:                  # [ppc64le]
  - 8                   # [ppc64le]

cdt_name:  # [linux]
  - cos7   # [ppc64le]

docker_image:                                           # [os.environ.get("BUILD_PLATFORM", "").startswith("linux")]
   - quay.io/condaforge/linux-anvil-ppc64le-cuda:11.2   # [ppc64le and os.environ.get("BUILD_PLATFORM") == "linux-ppc64le"]
```

leofang commented:
Should we move on? I propose to add CUDA 11.0, 11.1, 11.2 for both aarch64/ppc64le.

leofang commented:
Hi Isuru, you asked us for feedbacks 10 days ago and I'm trying to help move forward. If you already have your mind set we shouldn't be stuck here. Please feel free to do what you think is the best for CF and ignore me.

isuruf commented:
My mind is not set. I'm just trying to figure out if there's a need for < 11.2 or not. If there's no need, I'd like to go with just 11.2.

kkraus14 commented:
I'd propose we just ship 11.2 for now and if someone comes by and needs something else they can always raise an issue or a PR and we can expand later.

leofang commented on recipe/migrations/cuda_112_ppc64le_aarch64.yaml:
```diff
@@ -0,0 +1,37 @@
+  - 11.2                       # [ppc64le or aarch64]
+
+cudnn:                  # [ppc64le or aarch64]
+  - 8                   # [ppc64le or aarch64]
```
Can we set this to undefined for now? The PR to support both platforms is still blocked: https://github.com/conda-forge/cudnn-feedstock/pull/38

isuruf replied:
How is that relevant? The bot will stop at cudnn PR.

leofang replied:
If on these platforms cudnn is not yet available, would the migrator work (in non cudnn feedstocks)?

isuruf replied:
yes

prashantkhoje replied:
When is this migrator supposed to run? I don't see ppc64le build for ptxcompiler.

leofang replied:
@kkraus14 and I saw that manually applying the migrator would fail, likely because of a conflict with the global pinning (not sure): https://github.com/conda-forge/cub-feedstock/pull/17#issuecomment-1082005780

jakirkham replied:
Probably best to move any additional discussion into a new issue

leofang replied:
See #2679.
